pandas/core/frame.py

```diff
@@ -5400,8 +5402,13 @@ def _homogenize(data, index, dtype=None):
                 v = v.reindex(index, copy=False)
         else:
             if isinstance(v, dict):
-                v = _dict_compat(v)
-                oindex = index.astype('O')
+                if oindex is None:
+                    oindex = index.astype('O')
+
+                if isinstance(index, (DatetimeIndex, TimedeltaIndex)):
```

Comment:
This is to handle the case where the dict had `datetime/timedelta64`s as the keys - they're boxed in creating the index, so need to be changed in the dict too so the lookups work.  From this PR https://github.com/pydata/pandas/pull/10269.

AFAIK not needed for `PeriodIndex` since there aren't any types that get coerced to it?  This test covers nested construction with a `PeriodIndex`.
https://github.com/pydata/pandas/blob/master/pandas/tseries/tests/test_period.py#L2078